### PR TITLE
feat(npm): use install-tool instead of npm global installs

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

Uses install-tool for npm, yarn and pnpm instead of npm global installs.

## Context:

Preparation for #11984 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/npm/post-update/lerna.ts [hunk 1]
```diff
@@ -40,24 +41,21 @@ export async function generateLockFiles(
   const cmd = [];
   let cmdOptions = '';
   try {
+    const toolConstraints: ToolConstraint[] = [];
     if (lernaClient === 'yarn') {
-      let installYarn = 'npm i -g yarn';
-      const yarnCompatibility = config.constraints?.yarn;
-      if (validRange(yarnCompatibility)) {
```
Comment (rarkins) [inline lib/manager/npm/post-update/lerna.ts:46]: validity of constraints is now checked in exec

File: lib/manager/npm/post-update/lerna.ts [hunk 2]
```diff
@@ -40,24 +41,21 @@ export async function generateLockFiles(
...
         preCommands.push(getOptimizeCommand());
       }
       cmdOptions = '--ignore-scripts --ignore-engines --ignore-platform';
     } else if (lernaClient === 'npm') {
-      let installNpm = 'npm i -g npm';
-      const npmCompatibility = config.constraints?.npm;
-      if (validRange(npmCompatibility)) {
-        installNpm += `@${quote(npmCompatibility)} || true`;
-      }
-      preCommands.push(installNpm, 'hash -d npm');
```
Comment (rarkins) [inline lib/manager/npm/post-update/lerna.ts:60]: Might still need to do `hash -d npm`

File: lib/manager/npm/post-update/pnpm.ts [hunk 3]
```diff
@@ -23,30 +22,31 @@ export async function generateLockFile(
...
     const pnpmCompatibility = pnpmUpdate
       ? pnpmUpdate.newValue
       : config.constraints?.pnpm;
-    if (validRange(pnpmCompatibility)) {
-      installPnpm += `@${quote(pnpmCompatibility)}`;
-    }
-    const preCommands = [installPnpm];
+    const toolConstraints: ToolConstraint[] = [
+      {
+        toolName: 'npm',
```
Comment (rarkins) [inline lib/manager/npm/post-update/pnpm.ts:34]:
```suggestion
        toolName: 'pnpm',
```

File: lib/manager/npm/post-update/yarn.ts [hunk 4]
```diff
@@ -81,21 +82,21 @@ export async function generateLockFile(
...
       minYarnVersion && gte(minYarnVersion, '2.2.0');
     const isYarnModeAvailable = minYarnVersion && gte(minYarnVersion, '3.0.0');
-
-    let installYarn = 'npm i -g yarn';
+    const toolConstraint: ToolConstraint = {
+      toolName: 'yarn',
+    };
     if (isYarn1 && minYarnVersion) {
-      installYarn += `@${quote(yarnCompatibility)}`;
+      toolConstraint.constraint = yarnCompatibility;
```
Comment (rarkins) [inline lib/manager/npm/post-update/yarn.ts:89]: The Yarn 1 vs 2/3 works is non-standard and we should double check that this logic still holds. If this `if` is false then we just install latest `yarn`, which is currently v1

File: lib/manager/npm/post-update/yarn.ts [hunk 5]
```diff
@@ -81,21 +82,21 @@ export async function generateLockFile(
...
       NPM_CONFIG_CACHE: env.NPM_CONFIG_CACHE,
       npm_config_store: env.npm_config_store,
       CI: 'true',
     };
 
     const commands = [];
+    const preCommands = [];
     let cmdOptions = ''; // should have a leading space
     if (config.skipInstalls !== false) {
       if (isYarn1) {
```
Comment (rarkins) [inline lib/manager/npm/post-update/yarn.ts:102]: We need to check that `yarn` is still installed in the same place and that the `sed` logic below this works

File: lib/manager/npm/post-update/lerna.ts [hunk 6]
```diff
@@ -1,9 +1,10 @@
-import semver, { validRange } from 'semver';
+import semver from 'semver';
 import { quote } from 'shlex';
 import { getGlobalConfig } from '../../../config/global';
 import { TEMPORARY_ERROR } from '../../../constants/error-messages';
 import { logger } from '../../../logger';
 import { ExecOptions, exec } from '../../../util/exec';
+import { ToolConstraint } from '../../../util/exec/types';
```
Comment (viceice) [inline lib/manager/npm/post-update/lerna.ts:7]:
```suggestion
import type { ToolConstraint } from '../../../util/exec/types';
```

File: lib/manager/npm/post-update/lerna.ts (hunk 2 above)
Comment (viceice) [reply lib/manager/npm/post-update/lerna.ts:60]: yes, maybe i can fix it at buildpack before we merge this

File: lib/manager/npm/post-update/yarn.ts (hunk 4 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:89]: yes, we still need a global yarn v1

File: lib/manager/npm/post-update/yarn.ts (hunk 5 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:102]: yes, the yarn patch can be complicated. my original idea in buildpack was to not install those npm packages to normal global path.

Instead i would like to install to a separate folder like `/usrc/local/yarn/x.y.z` and add the wrapper.
When installing node, i would always create the npm wrapper, so the global npm always points to right version (latest installed).

With seperate folders, we can have multiple versions of those installed at same time and only need to switch the path.

so not really sure how to fix this. maybe add a helper script to buildpack, which does the yarn hack?

File: lib/manager/npm/post-update/yarn.ts (hunk 4 above)
Comment (rarkins) [reply lib/manager/npm/post-update/yarn.ts:89]: Sounds like our logic is still ok here

File: lib/manager/npm/post-update/yarn.ts (hunk 5 above)
Comment (rarkins) [reply lib/manager/npm/post-update/yarn.ts:102]: Another challenge is that there's theoretically the possibility - once we move to binarySource=install that one repo uses slim yarn and the next needs full install. So maybe we should have the concept of "install-tool yarn-slim" and be ready to maintain both patched and non patched versions side by side? Whichever is the latest to be installed then becomes "yarn". Renovate decides which of the two to install each time.

Comment (viceice) [reply lib/manager/npm/post-update/yarn.ts:102]: That Sounds good. Will be more easy to implement. What about repo local yarn path installs?

Comment (rarkins) [reply lib/manager/npm/post-update/yarn.ts:102]: Those are a temporary clone so ok to repeat it in those

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**